delay_distributions.Rmd

```diff
@@ -0,0 +1,206 @@
+One way that this could look, for examle, is:
+
+```{r delay_model}
+library("cmdstanr")
+
+delay_model <- '
+  data {
+    int n; // number of data points (rows with hospiatlisation date)
+    array[n] real onset_to_hosp; // onset to hospitalisation delays in the data
+  }
+
+  parameters {
+     real meanlog; // parameter to be estimtated
+     real<lower = 0> sdlog; // parameter to be estimated
+  }
+
+  model {
+    meanlog ~ normal(0, 10); // prior distribution
+    sdlog ~ normal(0, 10) T[0, ]; // prior distribution
+
+    onset_to_hosp ~ lognormal(meanlog, sdlog); // likelihood
+  }
+'
+```
+
+** Take 5 minutes ** to try to understand the model code and the relationships it encodes.
+
+You can now sample from the posterior of this model, i.e. conduct inference, by first compiling the model code.
+
+```{r compile_delay_model}
+file <- write_stan_file(delay_model)
+model <- cmdstan_model(file)
+```
+
+Once this is done, you can _sample_ from the model's posterior distribution:
+
+```{r sample_delay_model}
+model$sample(
+  data = list(
+    n = nrow(na.omit(hosp_scaled)),
+    onset_to_hosp = na.omit(hosp_scaled)$onset_to_hosp
+  )
+```
+
+This will probably cause an error - can you think of a reason why (if you don't see an error you could still think about what may cause an error, or resimulate the data that is being fitted to by re-running the commands that add columns to the infection times data set)?
+
+Can you change something that will avoid the error?
+You can look at [our example](delay_distribution_solution.html#estimating-delay-distributions) for a suggested fix, before moving on to a better solution in the next section.
+
+# Accounting for censoring
+
+A better way to account for censoring is to integrate over a window:
+
+```r
+delay_model_discrete <- '
+  data {
+    int n;
+    array[n] real onset_to_hosp;
+  }
+
+  transformed data {
+    array[n] real onset_to_hosp_upper;
+    array[n] real onset_to_hosp_lower;
+
+    for (i in 1:n) {
+      onset_to_hosp_upper[i] = onset_to_hosp[i] + 1;
+      onset_to_hosp_lower[i] = onset_to_hosp[i] - 1;
+    }
+  }
+
+  parameters {
+     real meanlog;
+     real<lower = 0> sdlog;
+  }
+
+  model {
+    // priors
+    meanlog ~ normal(0, 10);
+    sdlog ~ normal(0, 10);
+
+    // likelihood
+    for (i in 1:n) {
+      if (onset_to_hosp_lower[i] > 0) {
+        target += log(
+          lognormal_cdf(onset_to_hosp_upper[i] | meanlog, sdlog) - 
+          lognormal_cdf(onset_to_hosp_lower[i] | meanlog, sdlog)
+        );
+      } else {
+        target += lognormal_lcdf(onset_to_hosp_upper[i] | meanlog, sdlog);
+      }
+    }
+  }
+'
+
+file <- write_stan_file(delay_model_discrete)
+model <- cmdstan_model(file)
+
+model$sample(
+  data = list(
+    n = nrow(na.omit(hosp_scaled)),
+    onset_to_hosp = na.omit(hosp_scaled)$onset_to_hosp
+  )
+)
+```
+
+You will see that this solution does not require any changes to the data in order to obtain an estimate. 
```

Comment:
Is the lecture going to cover some of the why as this just feels pulled out of thin air?

Reply:
I think ideally this session would be followed by an explanation why, i.e. do first then see the theory behind it.

Reply:
I see